include/mata/utils/partition-relation-pair.hh

```diff
@@ -0,0 +1,1396 @@
+ * When the matrix is extended, additional (size * 2) + 1 elements of the
+ * vector are allocated. The matrix is traversed in some kind of "cascading" way
+ * as follows:
+ *
+ * Each number in the matrix corresponds to the order of accessing that element
+ * using this "cascading traversal".
+ *
+ *   MATRIX:
+ *   -----------------
+ *   | 0 | 3 | 8 | 15|
+ *   -----------------
+ *   | 1 | 2 | 7 | 14|
+ *   -----------------
+ *   | 4 | 5 | 6 | 13|
+ *   -----------------
+ *   | 9 | 10| 11| 12|
+ *   -----------------
+ *
+ *   VECTOR:
+ *   -----------------------------------------------------------------------
+ *   | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
+ *   -----------------------------------------------------------------------
+ *
+ * The data cell matrix[i][j] could be accessed using the formula
+ * vector[i >= j ? i * i + j : j * j + 2 * j - i].
+ *   
+ * Using this approach, there is no need to allocate unnecessary data cells
+ * when extending n x n matrix to the (n+1) x (n+1) matrix (in contrast with
+ * using row or column traversal).
+ * Since 'capacity * capacity' data cells were reserved, the vector won't ever
+ * be moved in the memory due to the extending. However, it requieres to
+ * reserve a lot of memory space when the capacity is a huge number.
+ * The data locality won't be violated since all of the elements will be stored
+ * as a contiguous vector.
+ *
+**/
+template <typename T>
+class CascadeSquareMatrix : public ExtendableSquareMatrix<T> {
+    private:
+ 
+        // data are stored in a single vector
+        std::vector<T> data_{};
+    
+    public:
+    
+        // constructors
+        CascadeSquareMatrix(size_t max_rows, size_t init_rows = 0);
+        CascadeSquareMatrix(const CascadeSquareMatrix& other);
+        
+        // implemented virtual functions
+        void set(size_t i, size_t j, T value) override;
+        T get(size_t i, size_t j) const override;
+        void extend(T placeholder = T()) override;
+        
+        // cloning
+        CascadeSquareMatrix *clone(void) const { 
+            return new CascadeSquareMatrix<T>(*this); }
+            
+        // operators
+        CascadeSquareMatrix<T>& operator=(const CascadeSquareMatrix<T>& other);
+        
+}; // CascadeSquareMatrix
+
+/**
+* @brief creates a Cascade square matrix
+* @param max_rows capacity of the square matrix
+* @param init_rows initial size of the square matrix
+*/
+template <typename T>
+CascadeSquareMatrix<T>::CascadeSquareMatrix(
+    size_t max_rows, size_t init_rows) {
+    
+    assert(init_rows <= max_rows && 
+           "Initial size of the matrix cannot be bigger than the capacity");
+    
+    this->m_type = MatrixType::Cascade;
+    this->capacity_ = max_rows;
+    data_.reserve(this->capacity_ * this->capacity_);
+    
+    // creating the initial size and filling the data cells with
+    // default values
+    for(size_t i = 0; i < init_rows; ++i) {extend();}
```

Comment:
The `codecov` says that this particular line (and two other lines similar to this one) is not fully covered even though I have implemented several tests which cause the following behaviour:

- the cycle is not executed since the condition is not satisfied
- the cycle is executed once
- the cycle is executed multiple times

Any thoughts?